api/handler/put.go

```diff
@@ -14,6 +23,118 @@ import (
 	"go.uber.org/zap"
 )
 
+type postPolicy struct {
+	Expiration time.Time          `json:"expiration"`
+	Conditions []*policyCondition `json:"conditions"`
+	empty      bool
+}
+
+func (p *postPolicy) condition(key string) *policyCondition {
+	for _, condition := range p.Conditions {
+		if condition.Key == key {
+			return condition
+		}
+	}
+	return nil
+}
+
+func (p *postPolicy) CheckContentLength(size int64) bool {
+	if p.empty {
+		return true
+	}
+	for _, condition := range p.Conditions {
+		if condition.Matching == "content-length-range" {
+			length := strconv.FormatInt(size, 10)
+			return condition.Key <= length && length <= condition.Value
+		}
+	}
+	return true
+}
+
+func (p *policyCondition) match(value string) bool {
+	switch p.Matching {
+	case "eq":
+		p.Matched = p.Value == value
+	case "starts-with":
+		if p.Key == api.ContentType {
+			p.Matched = true
+			for _, contentType := range strings.Split(value, ",") {
+				if !strings.HasPrefix(contentType, p.Value) {
+					p.Matched = false
+				}
+			}
+		} else {
+			p.Matched = strings.HasPrefix(value, p.Value)
+		}
+	}
+	return p.Matched
+}
+
+func (p *postPolicy) CheckField(key string, value string) error {
+	if p.empty {
+		return nil
+	}
+	cond := p.condition(key)
+	if cond == nil {
+		return errors.GetAPIError(errors.ErrPostPolicyConditionInvalidFormat)
+	}
+
+	if !cond.match(value) {
+		return errors.GetAPIError(errors.ErrPostPolicyConditionInvalidFormat)
+	}
+
+	return nil
+}
+
+func (p *postPolicy) AllConditionMatched() bool {
+	for _, condition := range p.Conditions {
+		if !condition.Matched {
+			return false
+		}
+	}
+	return true
+}
+
+type policyCondition struct {
+	Matching string
+	Key      string
+	Value    string
+	Matched  bool
+}
+
+func (p *policyCondition) UnmarshalJSON(data []byte) error {
+	var v interface{}
+	if err := json.Unmarshal(data, &v); err != nil {
+		return err
+	}
+
+	switch v := v.(type) {
+	case []interface{}:
+		if len(v) != 3 {
+			return fmt.Errorf("invalid condition")
+		}
+		p.Matching = v[0].(string)
+		if p.Matching == "content-length-range" {
+			p.Key = strconv.FormatFloat(v[1].(float64), 'f', 0, 32)
+			p.Value = strconv.FormatFloat(v[2].(float64), 'f', 0, 32)
+		} else {
+			p.Key = strings.ToLower(strings.TrimPrefix(v[1].(string), "$"))
+			p.Value = v[2].(string)
```

Comment:
Are there any cases when type conversion will lead to panic? There are plenty of conversions, maybe it will be safer to do `p.Value, ok = v[2].(string)` everywhere, not sure.